File: apps/client/local_plugins/core/knowledge_native/ios/Classes/KnowledgeEngine.swift
import Foundation

/// md 文库（真源）：每条笔记一份 `<id>.md`，落盘于 rootPath。镜像 Kotlin MarkdownVault。
final class MarkdownVault {
    struct Note { let id: String; let frontmatter: [String: String]; let body: String }

    private let rootPath: String
    init(_ rootPath: String) { self.rootPath = rootPath }

    private func root() -> URL {
        let url = URL(fileURLWithPath: rootPath, isDirectory: true)
        try? FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        return url
    }
    private func file(_ id: String) -> URL { root().appendingPathComponent("\(id).md") }

    func put(_ note: Note) {
        var s = "---\n"
        for (k, v) in note.frontmatter { s += "\(k): \(v.replacingOccurrences(of: "\n", with: " "))\n" }
        s += "---\n" + note.body
        try? s.write(to: file(note.id), atomically: true, encoding: .utf8)
    }

    func read(_ id: String) -> Note? {
        guard let raw = try? String(contentsOf: file(id), encoding: .utf8) else { return nil }
        return parse(id, raw)
    }

    func listIds() -> [String] {
        let items = (try? FileManager.default.contentsOfDirectory(atPath: root().path)) ?? []
        return items.filter { $0.hasSuffix(".md") }.map { String($0.dropLast(3)) }.sorted()
    }

    func exists(_ id: String) -> Bool { FileManager.default.fileExists(atPath: file(id).path) }
    func delete(_ id: String) { try? FileManager.default.removeItem(at: file(id)) }

    private func parse(_ id: String, _ raw: String) -> Note {
        var fm: [String: String] = [:]
        var body = raw
        if raw.hasPrefix("---") {
            let afterFence = raw.index(raw.startIndex, offsetBy: 3)
            if let endRange = raw.range(of: "\n---", range: afterFence..<raw.endIndex) {
                let block = String(raw[afterFence..<endRange.lowerBound])
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                for line in block.split(separator: "\n") {
                    if let ci = line.firstIndex(of: ":") {
                        let key = String(line[line.startIndex..<ci]).trimmingCharacters(in: .whitespaces)
                        let val = String(line[line.index(after: ci)...]).trimmingCharacters(in: .whitespaces)
                        fm[key] = val
                    }
                }
                if let nl = raw.range(of: "\n", range: endRange.upperBound..<raw.endIndex) {
                    body = String(raw[nl.upperBound...])
                } else {
                    body = ""
                }
            }
        }
        return Note(id: id, frontmatter: fm, body: body)
    }
}

/// 本地用户级知识库服务（原生门面）。镜像 Kotlin KnowledgeEngine。
///
/// 纯原生 agents_server 直接实例化调用（后台/设备唤醒，不经 Flutter）：
/// ```swift
/// let kb = KnowledgeEngine(vaultPath: dir)
/// kb.ingestMarkdown(userId: "u1", title: "连接蓝牙耳机", markdown: "…")
/// let ctx = kb.answerContext("蓝牙耳机怎么配对", userId: "u1")  // → 交 LLM
/// ```
final class KnowledgeEngine {
    private let vault: MarkdownVault
    private let embedder: EmbeddingProvider
    private let store: KnowledgeStore
    private let vectors: VectorStore
    private let chunker: Chunker
    private let retriever: Retriever
    private var counter = 0

    init(vaultPath: String,
         embedder: EmbeddingProvider = HashingEmbedding(),
         store: KnowledgeStore = InMemoryKnowledgeStore(),
         vectors: VectorStore = InMemoryVectorStore(),
         chunker: Chunker = Chunker()) {
        self.vault = MarkdownVault(vaultPath)
        self.embedder = embedder
        self.store = store
        self.vectors = vectors
        self.chunker = chunker
        self.retriever = Retriever(vectors: vectors, store: store, embedder: embedder)
    }

    private func nextId(_ prefix: String) -> String {
        let t = DispatchTime.now().uptimeNanoseconds
        let s = "\(prefix)-\(String(t, radix: 36))-\(String(counter, radix: 36))"
        counter += 1
        return s
    }

    @discardableResult
    func ingestMarkdown(userId: String, title: String, markdown: String,
                        kind: String = "note", sourceUri: String? = nil, id: String? = nil) -> String {
        let docId = id ?? nextId("doc")
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        var fm = ["id": docId, "title": title, "userId": userId, "kind": kind, "createdAt": "\(now)"]
        if let u = sourceUri { fm["uri"] = u }
        vault.put(MarkdownVault.Note(id: docId, frontmatter: fm, body: markdown))
        let doc = KbDocument(id: docId, sourceId: nextId("src"), userId: userId, title: title, createdAt: now)
        store.putDocument(doc)
        indexDocument(doc, markdown)
        return docId
    }

    func retrieve(_ query: String, k: Int = 6, userId: String? = nil) -> [ScoredChunk] {
        retriever.retrieve(query, k: k, userId: userId)
    }

    func answerContext(_ query: String, k: Int = 6, userId: String? = nil) -> String {
        let hits = retrieve(query, k: k, userId: userId)
        var s = ""
        for (i, h) in hits.enumerated() {
            s += "[\(i + 1)] 《\(h.documentTitle ?? h.chunk.documentId)》\n"
            s += h.chunk.text.trimmingCharacters(in: .whitespacesAndNewlines) + "\n\n"
        }
        return s.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func deleteDocument(_ docId: String) {
        vault.delete(docId)
        vectors.deleteByDocument(docId)
        store.deleteDocument(docId)
    }

    @discardableResult
    func reindexFromVault(userId: String? = nil) -> Int {
        store.clearIndex(); vectors.clear()
        var n = 0
        for id in vault.listIds() {
            guard let note = vault.read(id) else { continue }
            let uid = note.frontmatter["userId"] ?? "default"
            if let want = userId, uid != want { continue }
            let createdAt = Int64(note.frontmatter["createdAt"] ?? "") ?? 0
            let doc = KbDocument(id: id, sourceId: "src-\(id)", userId: uid,
                                 title: note.frontmatter["title"], createdAt: createdAt)
            store.putDocument(doc)
            indexDocument(doc, note.body)
            n += 1
        }
        return n
    }

    func vectorCount() -> Int { vectors.count() }

    private func indexDocument(_ doc: KbDocument, _ markdown: String) {
        let pieces = chunker.chunk(markdown)
        var chunks: [KbChunk] = []
        for (i, piece) in pieces.enumerated() {
            chunks.append(KbChunk(id: "\(doc.id)#\(i)", documentId: doc.id, userId: doc.userId,
                                  ordinal: i, text: piece.0, loc: piece.1))
        }
        store.putChunks(chunks)
        let vecs = embedder.embedBatch(chunks.map { $0.text })
        var recs: [VectorRecord] = []
        for (i, c) in chunks.enumerated() {
            recs.append(VectorRecord(chunkId: c.id, documentId: doc.id, userId: doc.userId, vector: vecs[i]))
        }
        vectors.upsert(recs)
    }
}

File: apps/client/local_plugins/core/knowledge_native/android/src/main/kotlin/ai/unihelper/knowledge/KnowledgeEngine.kt
package ai.unihelper.knowledge

import java.io.File

/**
 * md 文库（真源）：每条笔记一份 `<id>.md`，落盘于 [rootPath]。
 */
class MarkdownVault(private val rootPath: String) {
    data class Note(val id: String, val frontmatter: Map<String, String>, val body: String)

    private fun root(): File = File(rootPath).apply { if (!exists()) mkdirs() }
    private fun file(id: String): File = File(root(), "$id.md")

    fun put(note: Note) {
        val sb = StringBuilder()
        sb.append("---\n")
        for ((k, v) in note.frontmatter) sb.append("$k: ${v.replace("\n", " ")}\n")
        sb.append("---\n")
        sb.append(note.body)
        file(note.id).writeText(sb.toString())
    }

    fun read(id: String): Note? {
        val f = file(id)
        if (!f.exists()) return null
        return parse(id, f.readText())
    }

    fun listIds(): List<String> =
        root().listFiles()?.filter { it.isFile && it.name.endsWith(".md") }
            ?.map { it.name.removeSuffix(".md") }?.sorted() ?: emptyList()

    fun exists(id: String): Boolean = file(id).exists()

    fun delete(id: String) { file(id).let { if (it.exists()) it.delete() } }

    private fun parse(id: String, raw: String): Note {
        val fm = LinkedHashMap<String, String>()
        var body = raw
        if (raw.startsWith("---")) {
            val end = raw.indexOf("\n---", 3)
            if (end != -1) {
                for (line in raw.substring(3, end).trim().split("\n")) {
                    val i = line.indexOf(':')
                    if (i > 0) fm[line.substring(0, i).trim()] = line.substring(i + 1).trim()
                }
                val nl = raw.indexOf('\n', end + 1)
                body = if (nl == -1) "" else raw.substring(nl + 1)
            }
        }
        return Note(id, fm, body)
    }
}

/**
 * 本地用户级知识库服务（原生门面）。
 *
 * 纯原生 `agents_server` 直接实例化并调用（后台/设备唤醒场景，不经 Flutter）：
 * ```kotlin
 * val kb = KnowledgeEngine(context.filesDir.resolve("kb_vault").path)
 * kb.ingestMarkdown("u1", "连接蓝牙耳机", "...")
 * val ctx = kb.answerContext("蓝牙耳机怎么配对", userId = "u1")   // → 交 LLM 生成答案
 * ```
 * 默认零外部依赖（内存索引 + 特征哈希嵌入）；平台层可换 sqlite-vec / ONNX 端口实现。
 */
class KnowledgeEngine(
    vaultPath: String,
    private val embedder: EmbeddingProvider = HashingEmbedding(),
    private val store: KnowledgeStore = InMemoryKnowledgeStore(),
    private val vectors: VectorStore = InMemoryVectorStore(),
    private val chunker: Chunker = Chunker(),
) {
    private val vault = MarkdownVault(vaultPath)
    private val retriever = Retriever(vectors, store, embedder)
    private var counter = 0

    private fun nextId(prefix: String): String {
        val t = System.nanoTime()
        return "$prefix-${t.toString(36)}-${(counter++).toString(36)}"
    }

    /** 入库 markdown：写 md 文库(真源) + 建索引。返回 documentId。 */
    fun ingestMarkdown(
        userId: String,
        title: String,
        markdown: String,
        kind: String = "note",
        sourceUri: String? = null,
        id: String? = null,
    ): String {
        val docId = id ?: nextId("doc")
        val now = System.currentTimeMillis()
        val fm = linkedMapOf(
            "id" to docId, "title" to title, "userId" to userId,
            "kind" to kind, "createdAt" to now.toString(),
        )
        if (sourceUri != null) fm["uri"] = sourceUri
        vault.put(MarkdownVault.Note(docId, fm, markdown))
        val doc = KbDocument(docId, nextId("src"), userId, title, now)
        store.putDocument(doc)
        indexDocument(doc, markdown)
        return docId
    }

    /** 检索：混合(向量 + 关键词) → 带引用命中。 */
    fun retrieve(query: String, k: Int = 6, userId: String? = null): List<ScoredChunk> =
        retriever.retrieve(query, k, userId)

    /** 供 LLM/管家：检索 + 拼装带 [n] 引用的上下文（答案生成由调用方完成）。 */
    fun answerContext(query: String, k: Int = 6, userId: String? = null): String {
        val hits = retrieve(query, k, userId)
        val sb = StringBuilder()
        hits.forEachIndexed { i, h ->
            sb.append("[${i + 1}] 《${h.documentTitle ?: h.chunk.documentId}》\n")
            sb.append(h.chunk.text.trim()).append("\n\n")
        }
        return sb.toString().trimEnd()
    }

    fun deleteDocument(docId: String) {
        vault.delete(docId)
        vectors.deleteByDocument(docId)
        store.deleteDocument(docId)
    }

    /** 从 md 文库整库重建索引（证明索引派生自真源、可重建）。返回文档数。 */
    fun reindexFromVault(userId: String? = null): Int {
        store.clearIndex(); vectors.clear()
        var n = 0
        for (id in vault.listIds()) {
            val note = vault.read(id) ?: continue
            val uid = note.frontmatter["userId"] ?: "default"
            if (userId != null && uid != userId) continue
            val createdAt = note.frontmatter["createdAt"]?.toLongOrNull() ?: 0L
            val doc = KbDocument(id, "src-$id", uid, note.frontmatter["title"], createdAt)
            store.putDocument(doc)
            indexDocument(doc, note.body)
            n++
        }
        return n
    }

    fun vectorCount(): Int = vectors.count()

    private fun indexDocument(doc: KbDocument, markdown: String) {
        val pieces = chunker.chunk(markdown)
        val chunks = pieces.mapIndexed { i, (text, loc) ->
            KbChunk("${doc.id}#$i", doc.id, doc.userId, i, text, loc)
        }
        store.putChunks(chunks)
        val vecs = embedder.embedBatch(chunks.map { it.text })
        vectors.upsert(chunks.mapIndexed { i, c ->
            VectorRecord(c.id, doc.id, doc.userId, vecs[i])
        })
    }
}
